### DISCUSSION TITLE
DOM element measurements and lifecycle

### DISCUSSION THREAD

**Discussion by viridia** [solidjs/solid#1282] [Q&A] [answered]
So, I'm almost finished implementing a new Slider widget, except that there's a problem: in order to calculate the position of the thumb, I need to know the pixel width of the widget. This is easy enough to do once the component has been rendered, just take the 'ref' and get `ref.offsetWidth`. However, the problem comes when the widget first renders.

Apologies in advance for any misunderstandings - I'm trying to document my confusion :)

## First (naive) approach: 

```ts
style={{ left: `${position() * ref.offsetWidth - adjustment}px` }}
```

This doesn't work because `ref` is not defined when the component first renders and so we get an NPE.

## Second attempt:

```ts
style={ref && { left: `${position() * ref.offsetWidth - adjustment}px` }}
```

Because `ref` is reactive, this will update the style attribute when the ref is filled in. Unfortunately, `offsetWidth` is the wrong value when this happens, because layout hasn't happened yet. The formula won't get recalculated after layout because `offsetWidth` is a not a reactive variable.

## Third attempt:

```ts
const elementWidth = createMemo(() => {
  return ref?.offsetWidth ?? 0;
});
```

This suffers from the same problem, offsetWidth is not reactive. Also, I'm not sure whether `ref` is reactive when used in this way - I know it's reactive in JSX because of compiler magic, but I'm not sure if that applies to effects and memos. However, in this specific case, the memo doesn't get evaluated until after mount, so it kind of avoids being undefined by accident. OffsetWidth is still wrong, though.

## Fourth attempt:

OK let's look in solid-primitives - I see they have an element size function. Ooops, it says that "Target also takes falsy values to disable tracking." Unfortunately `ref` is going to be falsey initially, does that mean tracking is disabled forever?

## Fifth attempt:

Maybe a "use" directive? Let's look at the Solid API docs - reading over this section several times, I don't see how I can make this fit my use case. In fact, I'm having a hard time imagining what use case the "use" directive solves. The docs are fairly abstract here, and the toy example in the docs doesn't look much like real-world code. There's no explanation of what "model" or "accessor" are supposed to represent in the [docs for use](https://www.solidjs.com/docs/latest/api#use___).

## Sixth attempt:

OK so I guess I'm going to have to add a resize event listener myself - and make it so that it understands that the element is going to be undefined initially, only subscribing after `ref` has been filled in. Now, where to add that? It would be nice to make it something like a hook - except that if we pass `ref` to a hook, it won't be reactive inside the hook (since `ref` is not a function). Again, `ref` is only reactive inside the JSX block, as far as I can tell.

We could use the functional form of `ref` like this: `ref={createResizeListener}`, except once we do that we can't use `ref` for anything else.

Now, I know that I can solve this problem if I hack around it enough, but what I'm really getting at is that there's a surprising amount of friction involved with what I think should be a fairly straightforward task. That being said, I look forward to being told why I am wrong :)

BTW, just an aside - I noticed that there's a difference in behavior how React and Solid handle unmounting of references: in React, when a component is unmounted, the 'ref' callback gets called with `null`. That's not a huge deal, but is slightly surprising because I have (in the past) written React code that depended on that behavior.

**Comment by ryansolid** [comment, marked as answer]
A variable can never be reactive. Both 2 and 3 are not reactive.

If you want to react to a ref change. Use the callback form with a signal:

```js
const [ref, setRef] = createSignal()

return <div ref={setRef} style={ref() && { left: `${position() * ref().offsetWidth - adjustment}px` }} />
```
Does that work?

**Reply by viridia** [reply]
Thanks! Here's what I ended up doing:

```ts
interface ElementSize {
  width: number;
  height: number;
}

export const createElementSize = <T extends HTMLElement>(): [
  Accessor<ElementSize>,
  (element: T) => void
] => {
  const [box, setBox] = createSignal<ElementSize>({ width: 0, height: 0 });
  return [
    box,
    (element: T) => {
      if (!isServer) {
        new ResizeObserver(args => {
          const cr = args[0].contentRect;
          setBox({ width: cr.width, height: cr.height });
        }).observe(element);
      }
    },
  ];
};
```

```ts
const [box, sizeRef] = createElementSize();

return <div ref={forkRefs([sizeRef, otherRefs])} ... />
```
